ibis/spark/compiler.py

```diff
@@ -26,13 +68,356 @@ class SparkContext(ImpalaContext):
     pass
 
 
+def _is_inf(translator, expr):
+    arg, = expr.op().args
+    return '{} = {} or {} = {}'.format(
+        *map(
+            translator.translate,
+            [
+                arg,
+                ibis.literal(math.inf, type='double'),
+                arg,
+                ibis.literal(-math.inf, type='double'),
+            ]
+        )
+    )
+
+
+_sql_type_names = impala_compiler._sql_type_names.copy()
+
+_sql_type_names.update({
+    'date': 'date',
+})
+
+
+def _cast(translator, expr):
+    op = expr.op()
+    arg, target_type = op.args
+    arg_formatted = translator.translate(arg)
+
+    if isinstance(arg, ir.CategoryValue) and target_type == dt.int32:
+        return arg_formatted
+    if isinstance(arg, ir.TemporalValue) and target_type == dt.int64:
+        return '1000000 * unix_timestamp({})'.format(arg_formatted)
+    else:
+        sql_type = _type_to_sql_string(target_type)
+        return 'CAST({} AS {})'.format(arg_formatted, sql_type)
+
+
+def _type_to_sql_string(tval):
+    if isinstance(tval, dt.Decimal):
+        return 'decimal({}, {})'.format(tval.precision, tval.scale)
+    name = tval.name.lower()
+    try:
+        return _sql_type_names[name]
+    except KeyError:
+        raise com.UnsupportedBackendType(name)
+
+
+def _timestamp_op(func):
+    def _formatter(translator, expr):
+        op = expr.op()
+        left, right = op.args
+        formatted_left = translator.translate(left)
+        formatted_right = translator.translate(right)
+        func_name = func
+
+        if isinstance(left, (ir.TimestampScalar, ir.DateValue)):
+            formatted_left = 'cast({} as timestamp)'.format(formatted_left)
+
+        if isinstance(right, (ir.TimestampScalar, ir.DateValue)):
+            formatted_right = 'cast({} as timestamp)'.format(formatted_right)
+
+        # Spark only supports date_add, date_sub, and add_months.
+        # date_add and date_sub are in days, add_months is in months.
+        # These ops set HH:MM:SS to zero, operating only on date.
+        #
+        # If we use the <timestamp> + INTERVAL <int> <unit> syntax,
+        # Spark only supports the second argument being a literal.
+        if func_name in ('date_add', 'date_sub'):
```

Comment:
Can we break this logic out into functions that encapsulate the various cases that are being handled here? It's hard to tell by reading the code which case is which.

Reply:
Moving to new branch.